File: apps/android/app/src/main/java/com/litter/android/state/AppModel.kt
package com.litter.android.state

import com.litter.android.core.bridge.UniffiInit
import com.litter.android.util.LLog
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.atomic.AtomicLong
import uniffi.codex_mobile_client.AppServerRpc
import uniffi.codex_mobile_client.AppSnapshotRecord
import uniffi.codex_mobile_client.AppStore
import uniffi.codex_mobile_client.AppStoreSubscription
import uniffi.codex_mobile_client.AppStoreUpdateRecord
import uniffi.codex_mobile_client.DiscoveryBridge
import uniffi.codex_mobile_client.HandoffManager
import uniffi.codex_mobile_client.MessageParser
import uniffi.codex_mobile_client.ModelListParams
import uniffi.codex_mobile_client.ServerBridge
import uniffi.codex_mobile_client.SshBridge
import uniffi.codex_mobile_client.ThreadKey
import uniffi.codex_mobile_client.ThreadListParams

/**
 * Central app state singleton. Thin wrapper over Rust [AppStore] — all business
 * logic, reconciliation, and state management lives in Rust.
 *
 * Exposes a [snapshot] StateFlow that the UI observes. Updated automatically
 * via the Rust subscription stream.
 */
class AppModel private constructor(context: android.content.Context) {

    data class ComposerPrefillRequest(
        val requestId: Long,
        val threadKey: ThreadKey,
        val text: String,
    )

    companion object {
        private var _instance: AppModel? = null

        val shared: AppModel
            get() = _instance ?: throw IllegalStateException("AppModel not initialized — call init(context) first")

        fun init(context: android.content.Context): AppModel {
            if (_instance == null) {
                _instance = AppModel(context.applicationContext)
            }
            return _instance!!
        }
    }

    // --- Rust bridges (singletons behind the scenes) -------------------------

    val store: AppStore
    val rpc: AppServerRpc
    val discovery: DiscoveryBridge
    val serverBridge: ServerBridge
    val ssh: SshBridge
    val sshSessionStore: SshSessionStore
    val parser: MessageParser
    val launchState: AppLaunchState
    val appContext: android.content.Context = context

    init {
        UniffiInit.ensure(context)
        LLog.bootstrap(context)
        store = AppStore()
        rpc = AppServerRpc()
        discovery = DiscoveryBridge()
        serverBridge = ServerBridge()
        ssh = SshBridge()
        sshSessionStore = SshSessionStore(ssh)
        parser = MessageParser()
        launchState = AppLaunchState(context)
    }

    // --- Observable state ----------------------------------------------------

    private val _snapshot = MutableStateFlow<AppSnapshotRecord?>(null)
    val snapshot: StateFlow<AppSnapshotRecord?> = _snapshot.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()
    private val loadingModelServerIds = mutableSetOf<String>()
    private val loadingRateLimitServerIds = mutableSetOf<String>()
    private val sessionListMutex = Mutex()

    // --- Composer prefill queue (for edit message / slash commands) -----------

    private val nextComposerPrefillRequestId = AtomicLong(0)
    private val _composerPrefillRequest = MutableStateFlow<ComposerPrefillRequest?>(null)
    val composerPrefillRequest: StateFlow<ComposerPrefillRequest?> = _composerPrefillRequest.asStateFlow()

    fun queueComposerPrefill(threadKey: ThreadKey, text: String) {
        _composerPrefillRequest.value = ComposerPrefillRequest(
            requestId = nextComposerPrefillRequestId.incrementAndGet(),
            threadKey = threadKey,
            text = text,
        )
    }

    fun clearComposerPrefill(requestId: Long) {
        if (_composerPrefillRequest.value?.requestId == requestId) {
            _composerPrefillRequest.value = null
        }
    }

    // --- Subscription lifecycle ----------------------------------------------

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var subscriptionJob: Job? = null

    fun start() {
        if (subscriptionJob?.isActive == true) return
        subscriptionJob = scope.launch {
            try {
                refreshSnapshot()
                val subscription: AppStoreSubscription = store.subscribeUpdates()
                while (true) {
                    try {
                        val update: AppStoreUpdateRecord = subscription.nextUpdate()
                        LLog.d("AppModel", "AppStore update", fields = mapOf("update" to update::class.simpleName))
                        handleUpdate(update)
                    } catch (e: Exception) {
                        LLog.e("AppModel", "AppStore subscription loop failed", e)
                        throw e
                    }
                }
            } catch (e: Exception) {
                LLog.e("AppModel", "AppModel.start() subscription failed", e)
                _lastError.value = e.message
            }
        }
    }

    fun stop() {
        subscriptionJob?.cancel()
        subscriptionJob = null
    }

    // --- Snapshot refresh -----------------------------------------------------

    suspend fun refreshSnapshot() {
        try {
            val snap = applySavedServerNames(store.snapshot())
            _snapshot.value = snap
            _lastError.value = null
            val serverSummary = snap.servers.joinToString(separator = " | ") { server ->
                "${server.serverId}:${server.displayName}:${server.host}:${server.port}:${server.health}"
            }
            LLog.d(
                "AppModel",
                "snapshot refreshed",
                fields = mapOf("servers" to snap.servers.size, "summary" to serverSummary),
            )
        } catch (e: Exception) {
            _lastError.value = e.message
        }
    }

    private fun applySavedServerNames(snapshot: AppSnapshotRecord): AppSnapshotRecord {
        val nameByServerId = SavedServerStore.load(appContext)
            .mapNotNull { server ->
                val trimmed = server.name.trim()
                if (trimmed.isEmpty()) null else server.id to trimmed
            }
            .toMap()
        if (nameByServerId.isEmpty()) return snapshot

        return snapshot.copy(
            servers = snapshot.servers.map { server ->
                val savedName = nameByServerId[server.serverId]
                if (savedName != null && savedName != server.displayName) {
                    server.copy(displayName = savedName)
                } else {
                    server
                }
            },
            sessionSummaries = snapshot.sessionSummaries.map { summary ->
                val savedName = nameByServerId[summary.key.serverId]
                if (savedName != null && savedName != summary.serverDisplayName) {
                    summary.copy(serverDisplayName = savedName)
                } else {
                    summary
                }
            },
        )
    }

    suspend fun restartLocalServer() {
        val currentLocal = snapshot.value?.servers?.firstOrNull { it.isLocal }
        val serverId = currentLocal?.serverId ?: "local"
        val displayName = currentLocal?.displayName ?: "This Device"
        runCatching { serverBridge.disconnectServer(serverId) }
        serverBridge.connectLocalServer(
            serverId = serverId,
            displayName = displayName,
            host = "127.0.0.1",
            port = 0u,
        )
        restoreStoredLocalChatGptAuth(serverId)
        try {
            refreshSessions(listOf(serverId))
        } catch (_: Exception) {
        }
        refreshSnapshot()
    }

    suspend fun refreshSessions(serverIds: Collection<String>? = null) {
        val targetServerIds = (serverIds?.toList() ?: snapshot.value?.servers
            ?.filter { it.isConnected }
            ?.map { it.serverId }
            .orEmpty())
            .distinct()

        if (targetServerIds.isEmpty()) {
            return
        }

        sessionListMutex.withLock {
            try {
                for (serverId in targetServerIds) {
                    rpc.threadList(
                        serverId,
                        ThreadListParams(
                            cursor = null,
                            limit = null,
                            sortKey = null,
                            modelProviders = null,
                            sourceKinds = null,
                            archived = null,
                            cwd = null,
                            searchTerm = null,
                        ),
                    )
                }
                refreshSnapshot()
                _lastError.value = null
            } catch (e: Exception) {
                _lastError.value = e.message
                throw e
            }
        }
    }

    suspend fun loadConversationMetadataIfNeeded(serverId: String) {
        loadAvailableModelsIfNeeded(serverId)
        loadRateLimitsIfNeeded(serverId)
    }

    suspend fun loadAvailableModelsIfNeeded(serverId: String) {
        val server = snapshot.value?.servers?.firstOrNull { it.serverId == serverId } ?: return
        if (!server.isConnected) return
        if (server.availableModels != null) return
        if (!loadingModelServerIds.add(serverId)) return
        try {
            rpc.modelList(
                serverId,
                ModelListParams(cursor = null, limit = null, includeHidden = false),
            )
            refreshSnapshot()
        } catch (e: Exception) {
            _lastError.value = e.message
        } finally {
            loadingModelServerIds.remove(serverId)
        }
    }

    suspend fun loadRateLimitsIfNeeded(serverId: String) {
        val server = snapshot.value?.servers?.firstOrNull { it.serverId == serverId } ?: return
        if (!server.isConnected) return
        if (server.account == null) return
        if (server.rateLimits != null) return
        if (!loadingRateLimitServerIds.add(serverId)) return
        try {
            rpc.getAccountRateLimits(serverId)
            refreshSnapshot()
        } catch (e: Exception) {
            _lastError.value = e.message
        } finally {
            loadingRateLimitServerIds.remove(serverId)
        }
    }

    suspend fun restoreStoredLocalChatGptAuth(serverId: String) {
        val tokens = ChatGPTOAuthTokenStore(appContext).load() ?: return
        runCatching {
            rpc.loginAccount(
                serverId,
                uniffi.codex_mobile_client.LoginAccountParams.ChatgptAuthTokens(
                    accessToken = tokens.accessToken,
                    chatgptAccountId = tokens.accountId,
                    chatgptPlanType = tokens.planType,
                ),
            )
        }.onFailure { error ->
            _lastError.value = error.message
        }
    }

    suspend fun startTurn(
        key: ThreadKey,
        payload: AppComposerPayload,
    ) {
        try {
            store.startTurn(key, payload.toTurnStartParams(key.threadId))
            _lastError.value = null
        } catch (e: Exception) {
            _lastError.value = e.message
            throw e
        }
    }

    suspend fun externalResumeThread(
        key: ThreadKey,
        hostId: String? = null,
    ) {
        try {
            store.externalResumeThread(key, hostId)
            _lastError.value = null
        } catch (e: Exception) {
            _lastError.value = e.message
            throw e
        }
    }

    // --- Internal event handling ----------------------------------------------

    private suspend fun handleUpdate(update: AppStoreUpdateRecord) {
        // All update types trigger a snapshot refresh.
        // Rust's AppStore handles fine-grained state management internally.
        // We could optimize to only refresh affected parts, but snapshot()
        // is cheap since Rust builds it from in-memory state.
        refreshSnapshot()
    }
}

File: apps/ios/Sources/Litter/Models/AppModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class AppModel {
    /// Pre-built Rust objects initialized off the main thread to avoid
    /// priority inversion (tokio runtime init blocks at default QoS).
    private struct RustBridges: @unchecked Sendable {
        let store: AppStore
        let rpc: AppServerRpc
        let discovery: DiscoveryBridge
        let serverBridge: ServerBridge
        let ssh: SshBridge
    }

    /// Kick off Rust bridge construction on a background thread.
    /// Call from `AppDelegate.didFinishLaunching` before SwiftUI touches `shared`.
    nonisolated static func prewarmRustBridges() {
        _ = _prewarmResult
    }

    private nonisolated static let _prewarmResult: RustBridges = {
        RustBridges(
            store: AppStore(),
            rpc: AppServerRpc(),
            discovery: DiscoveryBridge(),
            serverBridge: ServerBridge(),
            ssh: SshBridge()
        )
    }()

    static let shared = AppModel()

    struct ComposerPrefillRequest: Identifiable, Equatable {
        let id = UUID()
        let threadKey: ThreadKey
        let text: String
    }

    let store: AppStore
    let rpc: AppServerRpc
    let discovery: DiscoveryBridge
    let serverBridge: ServerBridge
    let ssh: SshBridge

    private(set) var snapshot: AppSnapshotRecord?
    private(set) var lastError: String?
    private(set) var composerPrefillRequest: ComposerPrefillRequest?

    @ObservationIgnored private var subscription: AppStoreSubscription?
    @ObservationIgnored private var updateTask: Task<Void, Never>?
    @ObservationIgnored private var loadingModelServerIds: Set<String> = []

    init(
        store: AppStore? = nil,
        rpc: AppServerRpc? = nil,
        discovery: DiscoveryBridge? = nil,
        serverBridge: ServerBridge? = nil,
        ssh: SshBridge? = nil
    ) {
        let bridges = Self._prewarmResult
        self.store = store ?? bridges.store
        self.rpc = rpc ?? bridges.rpc
        self.discovery = discovery ?? bridges.discovery
        self.serverBridge = serverBridge ?? bridges.serverBridge
        self.ssh = ssh ?? bridges.ssh
    }

    deinit {
        updateTask?.cancel()
    }

    func start() {
        guard updateTask == nil else { return }
        subscription = store.subscribeUpdates()
        updateTask = Task { [weak self] in
            guard let self else { return }
            await self.refreshSnapshot()
            guard let subscription = self.subscription else { return }
            while !Task.isCancelled {
                do {
                    let update = try await subscription.nextUpdate()
                    await self.handleStoreUpdate(update)
                } catch {
                    if Task.isCancelled { break }
                    self.lastError = error.localizedDescription
                    break
                }
            }
        }
    }

    func stop() {
        updateTask?.cancel()
        updateTask = nil
        subscription = nil
    }

    func refreshSnapshot() async {
        do {
            applySnapshot(try await store.snapshot())
        } catch {
            lastError = error.localizedDescription
        }
    }

    func restartLocalServer() async throws {
        let currentLocal = snapshot?.servers.first(where: \.isLocal)
        let serverId = currentLocal?.serverId ?? "local"
        let displayName = currentLocal?.displayName ?? "This Device"
        serverBridge.disconnectServer(serverId: serverId)
        _ = try await serverBridge.connectLocalServer(
            serverId: serverId,
            displayName: displayName,
            host: "127.0.0.1",
            port: 0
        )
        await restoreStoredLocalChatGPTAuth(serverId: serverId)
        await refreshSnapshot()
    }

    func restoreStoredLocalChatGPTAuth(serverId: String) async {
        guard let tokens = (try? ChatGPTOAuthTokenStore.shared.load()) ?? nil else {
            return
        }

        do {
            _ = try await rpc.loginAccount(
                serverId: serverId,
                params: .chatgptAuthTokens(
                    accessToken: tokens.accessToken,
                    chatgptAccountId: tokens.accountID,
                    chatgptPlanType: tokens.planType
                )
            )
        } catch {
            lastError = error.localizedDescription
        }
    }

    func applySnapshot(_ snapshot: AppSnapshotRecord?) {
        self.snapshot = snapshot
        if snapshot != nil {
            lastError = nil
        }
    }

    private func handleStoreUpdate(_ update: AppStoreUpdateRecord) async {
        switch update {
        case .threadChanged(let key):
            await refreshThreadSnapshot(key: key)
        case .threadRemoved(let key):
            removeThreadSnapshot(for: key)
        case .activeThreadChanged(let key):
            updateActiveThread(key)
            if let key, snapshot?.threadSnapshot(for: key) == nil {
                await refreshThreadSnapshot(key: key)
            }
        case .pendingApprovalsChanged:
            await refreshSnapshot()
        case .pendingUserInputsChanged:
            await refreshSnapshot()
        case .serverChanged:
            await refreshSnapshot()
        case .serverRemoved:
            await refreshSnapshot()
        case .fullResync:
            await refreshSnapshot()
        case .voiceSessionChanged:
            await refreshSnapshot()
        case .realtimeTranscriptUpdated:
            break
        case .realtimeHandoffRequested:
            break
        case .realtimeSpeechStarted:
            break
        case .realtimeStarted:
            await refreshSnapshot()
        case .realtimeOutputAudioDelta:
            break
        case .realtimeError:
            await refreshSnapshot()
        case .realtimeClosed:
            await refreshSnapshot()
        }
    }

    private func refreshThreadSnapshot(key: ThreadKey) async {
        guard snapshot != nil else {
            await refreshSnapshot()
            return
        }

        do {
            guard let threadSnapshot = try await store.threadSnapshot(key: key) else {
                removeThreadSnapshot(for: key)
                return
            }
            applyThreadSnapshot(threadSnapshot)
        } catch {
            lastError = error.localizedDescription
            await refreshSnapshot()
        }
    }

    private func applyThreadSnapshot(_ thread: AppThreadSnapshot) {
        guard var snapshot else {
            applySnapshot(nil)
            return
        }

        if let index = snapshot.threads.firstIndex(where: { $0.key == thread.key }) {
            snapshot.threads[index] = thread
        } else {
            snapshot.threads.append(thread)
        }
        self.snapshot = snapshot
        lastError = nil
    }

    private func removeThreadSnapshot(for key: ThreadKey) {
        guard var snapshot else { return }
        snapshot.threads.removeAll { $0.key == key }
        if snapshot.activeThread == key {
            snapshot.activeThread = nil
        }
        self.snapshot = snapshot
    }

    private func updateActiveThread(_ key: ThreadKey?) {
        guard var snapshot else { return }
        snapshot.activeThread = key
        self.snapshot = snapshot
    }

    func queueComposerPrefill(threadKey: ThreadKey, text: String) {
        composerPrefillRequest = ComposerPrefillRequest(threadKey: threadKey, text: text)
    }

    func clearComposerPrefill(id: UUID) {
        guard composerPrefillRequest?.id == id else { return }
        composerPrefillRequest = nil
    }

    func availableModels(for serverId: String) -> [Model] {
        snapshot?.serverSnapshot(for: serverId)?.availableModels ?? []
    }

    func rateLimits(for serverId: String) -> RateLimitSnapshot? {
        snapshot?.serverSnapshot(for: serverId)?.rateLimits
    }

    func loadConversationMetadataIfNeeded(serverId: String) async {
        await loadAvailableModelsIfNeeded(serverId: serverId)
        await loadRateLimitsIfNeeded(serverId: serverId)
    }

    func loadAvailableModelsIfNeeded(serverId: String) async {
        guard let server = snapshot?.serverSnapshot(for: serverId), server.isConnected else { return }
        guard server.availableModels == nil else { return }
        guard !loadingModelServerIds.contains(serverId) else { return }
        loadingModelServerIds.insert(serverId)
        defer { loadingModelServerIds.remove(serverId) }
        do {
            _ = try await rpc.modelList(
                serverId: serverId,
                params: ModelListParams(cursor: nil, limit: nil, includeHidden: false)
            )
            await refreshSnapshot()
        } catch {
            lastError = error.localizedDescription
        }
    }

    func loadRateLimitsIfNeeded(serverId: String) async {
        guard let server = snapshot?.serverSnapshot(for: serverId), server.isConnected else { return }
        guard server.rateLimits == nil else { return }
        guard server.account != nil else { return }
        do {
            _ = try await rpc.getAccountRateLimits(serverId: serverId)
        } catch {
            lastError = error.localizedDescription
        }
    }

    func startTurn(key: ThreadKey, payload: AppComposerPayload) async throws {
        do {
            try await store.startTurn(
                key: key,
                params: payload.turnStartParams(threadId: key.threadId)
            )
        } catch {
            lastError = error.localizedDescription
            throw error
        }
    }

    func ensureThreadLoaded(
        key: ThreadKey,
        maxAttempts: Int = 5
    ) async -> ThreadKey? {
        if snapshot?.threadSnapshot(for: key) != nil {
            return key
        }

        var currentKey = key
        for attempt in 0..<maxAttempts {
            var readSucceeded = false
            do {
                let response = try await rpc.threadRead(
                    serverId: currentKey.serverId,
                    params: ThreadReadParams(
                        threadId: currentKey.threadId,
                        includeTurns: true
                    )
                )
                currentKey = ThreadKey(serverId: currentKey.serverId, threadId: response.thread.id)
                store.setActiveThread(key: currentKey)
                readSucceeded = true
            } catch {
                lastError = error.localizedDescription
            }

            await refreshSnapshot()
            if snapshot?.threadSnapshot(for: currentKey) != nil {
                return currentKey
            }

            if !readSucceeded {
                do {
                    _ = try await rpc.threadList(
                        serverId: currentKey.serverId,
                        params: ThreadListParams(
                            cursor: nil,
                            limit: nil,
                            sortKey: nil,
                            modelProviders: nil,
                            sourceKinds: nil,
                            archived: nil,
                            cwd: nil,
                            searchTerm: nil
                        )
                    )
                } catch {
                    lastError = error.localizedDescription
                }

                await refreshSnapshot()
                if snapshot?.threadSnapshot(for: currentKey) != nil {
                    return currentKey
                }
            }

            if attempt + 1 < maxAttempts {
                try? await Task.sleep(nanoseconds: 250_000_000)
            }
        }

        if let activeKey = snapshot?.activeThread,
           activeKey.serverId == currentKey.serverId,
           snapshot?.threadSnapshot(for: activeKey) != nil {
            return activeKey
        }

        return nil
    }
}

extension AppSnapshotRecord {
    func threadSnapshot(for key: ThreadKey) -> AppThreadSnapshot? {
        threads.first { $0.key == key }
    }

    func serverSnapshot(for serverId: String) -> AppServerSnapshot? {
        servers.first { $0.serverId == serverId }
    }

    func sessionSummary(for key: ThreadKey) -> AppSessionSummary? {
        sessionSummaries.first { $0.key == key }
    }

    func resolvedThreadKey(for receiverId: String, serverId: String) -> ThreadKey? {
        guard let normalized = AgentLabelFormatter.sanitized(receiverId) else { return nil }
        if let summary = sessionSummaries.first(where: {
            $0.key.serverId == serverId && $0.key.threadId == normalized
        }) {
            return summary.key
        }
        return ThreadKey(serverId: serverId, threadId: normalized)
    }

    func resolvedAgentTargetLabel(for target: String, serverId: String) -> String? {
        if AgentLabelFormatter.looksLikeDisplayLabel(target) {
            return AgentLabelFormatter.sanitized(target)
        }
        guard let normalized = AgentLabelFormatter.sanitized(target) else { return nil }
        if let summary = sessionSummaries.first(where: {
            $0.key.serverId == serverId && $0.key.threadId == normalized
        }) {
            return summary.agentDisplayLabel ?? AgentLabelFormatter.sanitized(target)
        }
        return nil
    }
}
